rasa/core/policies/ensemble.py

```diff
@@ -355,16 +355,42 @@ def is_not_memo_policy(best_policy_name) -> bool:
         )
         return not (is_memo or is_augmented)
 
-    def probabilities_using_best_policy(
+    @staticmethod
+    def is_not_mapping_policy(best_policy_name) -> bool:
+        from rasa.core.policies.mapping_policy import MappingPolicy
+
+        return not best_policy_name.endswith("_" + MappingPolicy.__name__)
+
+    def _best_policy_prediction(
         self, tracker: DialogueStateTracker, domain: Domain
     ) -> Tuple[Optional[List[float]], Optional[Text]]:
-        import numpy as np
+        """Picks the best policy prediction based on probabilities and policy priority.
+
+        Args:
+            tracker: the :class:`rasa.core.trackers.DialogueStateTracker`
+            domain: the :class:`rasa.core.domain.Domain`
+
+        Returns:
+            result: the list of probabilities for the next actions
+            best_policy_name: the name of the picked policy
+        """
+
+        from rasa.core.policies.form_policy import FormPolicy
 
         result = None
         max_confidence = -1
         best_policy_name = None
         best_policy_priority = -1
 
+        # form and mapping policies are special:
+        # form should be above fallback
+        # mapping should be below fallback
+        # mapping is above form if it wins over fallback
+        # therefore form predictions are stored separately
+        form_probabilities = None
+        form_policy_index = None
+        form_policy_priority = None
+
         for i, p in enumerate(self.policies):
             probabilities = p.predict_action_probabilities(tracker, domain)
```

Comment:
when we move this before the for loop and store the predictions in a dictionary, e.g. 

```
        predictions = {p.__class__.__name__: p.predict_action_probabilities(tracker, domain) for p in self.policies}
```

then we can make the code way shorter (contains some pseudo code and not working parts, but you'll get the idea). What do you think?

```python
        predictions = {p.__class__.__name__: p.predict_action_probabilities(tracker, domain) for p in self.policies}

        for i, (policy_name, probabilities) in enumerate(predictions):
            if policy_name == FormPolicy.__class__.__name__:
                continue
    
            if len(tracker.events) > 0 and isinstance(
                tracker.events[-1], ActionExecutionRejected
            ):
                probabilities[
                    domain.index_for_action(tracker.events[-1].action_name)
                ] = 0.0

            # pick the best policy
            confidence = max(probabilities)
            if (confidence, self.policies[i].priority) > (max_confidence, best_policy_priority):
                max_confidence = confidence
                result = probabilities
                best_policy_name = f"policy_{i}_{type(p).__name__}"
                best_policy_priority = p.priority

        if self._is_form_policy_present() is not None and self.is_not_mapping_policy(
            best_policy_name
        ):
            # if mapping didn't win, check form policy predictions
            if (max(predictions["FormPolicyName"]), form_policy_priority) > (
                max_confidence,
                best_policy_priority,
            ):
                result = form_probabilities
                best_policy_name = f"policy_{form_policy_index}_{FormPolicy.__name__}"

        return result, best_policy_name
```

Reply:
btw this part should evaluated outside of the loop for a bit better performance:

```
if len(tracker.events) > 0 and isinstance(
                tracker.events[-1], ActionExecutionRejected
            ):
``

Reply:
good idea, I'll do it